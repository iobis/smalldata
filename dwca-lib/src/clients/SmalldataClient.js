import { mapDatasetToRequest, mapOccurrenceToDwca } from './SmalldataConverters'

const authorizationValue = 'Basic verysecret'
const headers = {
  'Authorization': authorizationValue,
  'Content-Type':  'application/json'
}

export async function getDatasets() {
  const response = await fetch('/api/datasets', { headers })
    .then(response => response.json())
  return response.map(renameRefToId)
}

export async function createDataset(dataset) {
  const request = mapDatasetToRequest(dataset)
  return await fetch('/api/datasets', {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export function renameRefToId({ ref, ...rest }) {
  return ({ id: ref, ...rest })
}

export async function getOccurrences({ userRef }) {
  const query = 'projectFields=dwcRecord.tdwg.scientificName&projectFields=dwcRecord.tdwg.eventDate'
  return await fetch(`/api/dwca/user/${userRef}/records?${query}`, { headers })
    .then(response => response.json())
}

export async function getOccurrence({ datasetId, dwcaId, userRef }) {
  return fetch(`/api/dwca/${datasetId}/user/${userRef}/records/${encodeURIComponent(dwcaId)}`, { headers })
    .then(response => response.json())
}

export async function findLatestOccurrence({ userRef }) {
  const occurrences = await getOccurrences({ userRef })
  const latest = occurrences
    .filter(dwca => !!dwca.addedAtInstant)
    .sort((left, right) => left.addedAtInstant < right.addedAtInstant ? 1 : -1)[0]
  return await getOccurrence({ userRef, dwcaId: latest.dwcaId, datasetId: latest.dataset })
}

export async function updateOccurrence({ occurrence, userRef, dwcaId }) {
  const request = mapOccurrenceToDwca(occurrence)
  const url = `/api/dwca/${occurrence.dataset.id}/user/${userRef}/records/${encodeURIComponent(dwcaId)}`
  return await fetch(url, {
    method: 'PUT',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export async function createOccurrence({ occurrence, userRef }) {
  const request = mapOccurrenceToDwca(occurrence)
  const url = `/api/dwca/${occurrence.dataset.id}/user/${userRef}/records`
  return await fetch(url, {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
}

export async function getUsers() {
  return fetch('/api/users', { headers })
    .then(response => response.json())
}

export async function getUsersWithDatasets() {
  const [users, datasets] = await Promise.all([getUsers(), getDatasets()])
  const datasetIdToDataset = groupBy(datasets, 'id')
  return users.map(user => {
    if (!user['dataset_refs']) user['dataset_refs'] = []
    const datasets = user['dataset_refs'].map(datasetRef => datasetIdToDataset[datasetRef][0])
    return { ...user, datasets }
  })
}

export async function createUser({ datasetIds, email, name, role }) {
  const request = {
    'dataset_refs': datasetIds,
    emailAddress:   email,
    name,
    role
  }
  const url = '/api/users'
  return await fetch(url, {
    method: 'POST',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
    .catch(error => ({
      exception:        'Some error occurred',
      exceptionMessage: error
    }))
}

export async function updateUser({ id, datasetIds, email, name, role }) {
  const request = {
    'dataset_refs': datasetIds,
    emailAddress:   email,
    name,
    role
  }
  const url = '/api/users/' + id
  return await fetch(url, {
    method: 'PUT',
    headers,
    body:   JSON.stringify(request)
  }).then(response => response.json())
    .catch(error => ({
      exception:        'Some error occurred',
      exceptionMessage: error
    }))
}

function groupBy(list, props) {
  return list.reduce((a, b) => {
    (a[b[props]] = a[b[props]] || []).push(b)
    return a
  }, {})
}
