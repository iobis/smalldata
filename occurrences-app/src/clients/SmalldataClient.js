import { format } from 'date-fns'

export async function getDatasets() {
  const response = await fetch('/api/datasets')
    .then(response => response.json())
  return response.map(renameRefToId)
}

export function renameRefToId({ ref, ...rest }) {
  return ({ id: ref, ...rest })
}

export function datasetTitleOf(dataset) {
  const title = dataset && dataset.title && dataset.title.value
  return title || ''
}

export function getOccurrenceMock() {
  return [{
    id:             1,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }, {
    id:             2,
    addedDate:      '2011-12-01',
    scientificName: 'Abra Alba',
    dataset:        'NPPSD Short-tailed Albatross Sightings',
    occurrenceDate: '2011-12-09'
  }]
}

export async function postOccurrence({ occurence }) {
  const userRef = 'ovZTtaOJZ98xDDY'
  const occurrence = {
    core:       'occurrence',
    occurrence: [{
      tdwg: {
        datasetID:                     occurence.dataset.id,
        datasetName:                   occurence.dataset.title.value,
        basisOfRecord:                 occurence.occurrenceData.basisOfRecord.charAt(0).toUpperCase() + occurence.occurrenceData.basisOfRecord.slice(1),
        eventDate:                     format(occurence.occurrenceData.beginDate, 'YYYY-MM-DD') + '/' + format(occurence.occurrenceData.endDate, 'YYYY-MM-DD'),
        lifestage:                     occurence.occurrenceData.lifeStage,
        occurrenceStatus:              occurence.occurrenceData.occurrenceStatus,
        scientificName:                occurence.occurrenceData.scientificName,
        sex:                           occurence.occurrenceData.sex,
        decimalLongitude:              occurence.locationData.decimalLongitude,
        decimalLatitude:               occurence.locationData.decimalLatitude,
        coordinateUncertaintyInMeters: occurence.locationData.coordinateUncertainty,
        minimumDepthInMeters:          occurence.locationData.minimumDepth,
        maximumDepthInMeters:          occurence.locationData.maximumDepth,
        verbatimCoordinates:           occurence.locationData.verbatimCoordinates,
        verbatimDepth:                 occurence.locationData.verbatimDepth
      }
    }],
    emof:       [{
      purl:  {},
      iobis: {}
    }, {
      iobis: {}
    }]
  }
  const url = `/api/dwca/${occurence.dataset.id}/user/${userRef}/records`
  return await fetch(url, {
    method:  'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body:    JSON.stringify(occurrence)
  }).then(response => response.json())
}
