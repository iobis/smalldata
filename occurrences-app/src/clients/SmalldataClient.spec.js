import * as SmalldataClient from './SmalldataClient'
import { RESPONSE_DEFAULT } from './SmalldataClient.mock'

describe('SmalldataClient', () => {
  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(() =>
      new Promise((resolve) => {
        resolve({ json: () => RESPONSE_DEFAULT })
      })
    )
  })

  afterEach(() => {
    global.fetch.mockRestore()
  })

  it('getDatasets()', async() => {
    await SmalldataClient.getDatasets()
    expect(fetch).toHaveBeenCalledTimes(1)
    expect(fetch).toBeCalledWith('/api/datasets')
  })

  it('datasetTitleOf(dataset)', () => {
    expect(SmalldataClient.datasetTitleOf(RESPONSE_DEFAULT[0]))
      .toEqual('Caprellids polulation structure in Usujiri, Hokkaido, Japan')
    expect(SmalldataClient.datasetTitleOf(null))
      .toEqual('')
  })

  it('postOccurrence()', async() => {
    await SmalldataClient.postOccurrence({
      occurence: {
        dataset:        {
          id:    'wEaBfmFyQhYCdsk',
          title: {
            value: 'Caprellids polulation structure in Usujiri, Hokkaido, Japan'
          }
        },
        occurrenceData: {
          basisOfRecord:    'humanObservation',
          beginDate:        Date.UTC(2019, 3, 29),
          endDate:          Date.UTC(2019, 3, 30),
          lifeStage:        'adult',
          occurrenceStatus: 'present',
          scientificName:   'ala abra',
          sex:              'male'
        }
      }
    })

    expect(fetch).toHaveBeenCalledTimes(1)
    expect(fetch.mock.calls[0][0]).toBe('/api/dwca/wEaBfmFyQhYCdsk/user/ovZTtaOJZ98xDDY/records')
    expect(fetch.mock.calls[0][1].method).toBe('POST')
    expect(JSON.parse(fetch.mock.calls[0][1].body)).toEqual({
      core:       'occurrence',
      occurrence: [{
        tdwg: {
          datasetID:        'wEaBfmFyQhYCdsk',
          datasetName:      'Caprellids polulation structure in Usujiri, Hokkaido, Japan',
          basisOfRecord:    'HumanObservation',
          eventDate:        '2019-04-29/2019-04-30',
          lifestage:        'adult',
          occurrenceStatus: 'present',
          scientificName:   'ala abra',
          sex:              'male'
        }
      }],
      emof:       [{
        purl:  {},
        iobis: {}
      }, {
        iobis: {}
      }]
    })
  })
})
