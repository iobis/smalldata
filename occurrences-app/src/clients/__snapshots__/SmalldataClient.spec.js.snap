// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmalldataClient mapDwcaToMeasurements() 1`] = `
Array [
  Object {
    "type": "Abundance category of biological entity specified elsewhere",
    "unit": "Dimensionless",
    "units": Array [
      Object {
        "id": "http://vocab.nerc.ac.uk/collection/P06/current/UUUU/",
        "name": "Dimensionless",
      },
    ],
    "value": "3",
  },
  Object {
    "type": "Abundance of biological entity specified elsewhere per unit area of the bed",
    "unit": "Number per square meter",
    "units": Array [
      Object {
        "id": "http://vocab.nerc.ac.uk/collection/P06/current/UPMS/",
        "name": "Number per square meter",
      },
    ],
    "value": "44",
  },
  Object {
    "type": "ObservedIndividualCount",
    "unit": "Dimensionless",
    "units": Array [
      Object {
        "id": "http://vocab.nerc.ac.uk/collection/P06/current/UUUU/",
        "name": "Dimensionless",
      },
    ],
    "value": "1",
  },
  Object {
    "type": "Pressure",
    "unit": "Decibars",
    "units": Array [
      Object {
        "id": "http://vocab.nerc.ac.uk/collection/P06/current/UPDB/",
        "name": "Decibars",
      },
    ],
    "value": "1",
  },
  Object {
    "type": "Salinity",
    "unit": "Grams per kilogram",
    "units": Array [
      Object {
        "id": "http://vocab.nerc.ac.uk/collection/P06/current/UGKG",
        "name": "Grams per kilogram",
      },
      Object {
        "id": "http://vocab.nerc.ac.uk/collection/P06/current/UUUU/",
        "name": "PSU (dimensionless)",
      },
    ],
    "value": "3",
  },
  Object {
    "type": "Temperature",
    "unit": "Degrees Celsius",
    "units": Array [
      Object {
        "id": "http://vocab.nerc.ac.uk/collection/P06/current/UPAA",
        "name": "Degrees Celsius",
      },
    ],
    "value": "2",
  },
]
`;
